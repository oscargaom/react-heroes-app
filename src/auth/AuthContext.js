import { createContext } from "react";

/*  Al usar createContext este nos facilita información que puede ser 
    compartida con los demás componentes del árbol de componentes */
export const AuthContext = createContext();