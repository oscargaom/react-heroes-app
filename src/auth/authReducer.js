import { types } from "../types/types";

export const authReducer = (state = {}, action) => {
    
    // console.log('=========================')
    // console.log('authReducer');
    // console.log('state');
    // console.log(state);
    // console.log('action');
    // console.log(action);
    // console.log('=========================')
    
    switch (action?.type) {
        case types.login:
            return {
                ...action.payload, 
                logged: true
            }
        case types.logout:
            return {logged: false};
        default:
            return state;
    }
};
