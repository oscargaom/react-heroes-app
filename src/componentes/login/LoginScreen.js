import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

export const LoginScreen = ({ history }) => {


    const { authUser, authDispatch } = useContext(AuthContext);

    const handleLogin = () => {

        const user = {
            ...authUser,
            name: "Manuel"
        };
        
        const action = {
            type: types.login,
            payload: user
        }

        authDispatch(action);

        const lastPath = localStorage.getItem('lastPath') || "/";
        // console.log(lastPath);

        // history.push("/");

        /*  Tenemos acceso al history y otras variables por medio de las props las cuales 
            son creadas por el react-router-dom; en este caso replace cambia la historia 
            de navegación en LoginScreen impidiendo de este modo navegaciones que pueden
            tener comportamiento no deseados en nuestra navegación.
        */
        history.replace(lastPath);
    };

    return (
        <div className="container mt-5">
            <h1>Login</h1>
            <hr />

            <button
                className="btn btn-primary"
                onClick={handleLogin}
            >
                Login
            </button>
        </div>
    )
}
