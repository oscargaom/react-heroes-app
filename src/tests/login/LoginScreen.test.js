import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { LoginScreen } from '../../componentes/login/LoginScreen';
import { AuthContext } from '../../auth/AuthContext';
import { authContextLoggedValue } from '../fixtures/authContextValue';
import { historyMockBase } from '../fixtures/historyMock';
import { types } from '../../types/types';

describe('Pruebas en el componente <LoginScreen />', () => {

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('debe mostrarse en el snapshot ', () => {

        const wrapper = mount(
            <AuthContext.Provider value={authContextLoggedValue}>
                <LoginScreen history={historyMockBase} />
            </AuthContext.Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('debe modificar el localStorage', () => {
        
        const wrapper = mount(
            <AuthContext.Provider value={authContextLoggedValue}>
                <LoginScreen history={historyMockBase} />
            </AuthContext.Provider>
        );

        const handleClick = wrapper.find('button').prop('onClick'); 

        localStorage.setItem('lastPath', '/dc');
        handleClick();
        expect(historyMockBase.replace).toHaveBeenCalledWith('/dc');
    });

    test('debe realizar el dispatch y la navegaciÃ³n', () => {

        Storage.prototype.getItem = jest.fn();

        const wrapper = mount(
            <AuthContext.Provider value={authContextLoggedValue}>
                <LoginScreen history={historyMockBase} />
            </AuthContext.Provider>
        );

        // console.log(wrapper.html())
        wrapper.find('button').prop('onClick')();
        expect(authContextLoggedValue.authDispatch).toHaveBeenCalledTimes(1);
        expect(authContextLoggedValue.authDispatch).toHaveBeenCalledWith({
            type: types.login,
            payload: {
                name: 'Manuel',
                logged:true
            }
        });

        expect(historyMockBase.replace).toHaveBeenCalledTimes(1);
        expect(historyMockBase.replace).toHaveBeenCalledWith('/');

        expect(localStorage.getItem).toHaveBeenCalledTimes(1);
        expect(localStorage.getItem).toHaveBeenCalledWith('lastPath');
    });
})

