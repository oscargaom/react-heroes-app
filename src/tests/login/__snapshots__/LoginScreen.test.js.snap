// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <LoginScreen /> debe mostrarse en el snapshot  1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
